pkg/plugins/golang/v2/options.go

```diff
@@ -0,0 +1,250 @@
+	Namespaced bool
+
+	// Flags that define which parts should be scaffolded
+	DoAPI        bool
+	DoController bool
+	DoDefaulting bool
+	DoValidation bool
+	DoConversion bool
+}
+
+// Validate verifies that all the fields have valid values
+func (opts Options) Validate() error {
+	// Check that the required flags did not get a flag as their value
+	// We can safely look for a '-' as the first char as none of the fields accepts it
+	// NOTE: We must do this for all the required flags first or we may output the wrong
+	// error as flags may seem to be missing because Cobra assigned them to another flag.
+	if strings.HasPrefix(opts.Group, "-") {
+		return fmt.Errorf(groupPresent)
+	}
+	if strings.HasPrefix(opts.Version, "-") {
+		return fmt.Errorf(versionPresent)
+	}
+	if strings.HasPrefix(opts.Kind, "-") {
+		return fmt.Errorf(kindPresent)
+	}
+	// Now we can check that all the required flags are not empty
+	if len(opts.Group) == 0 {
+		return fmt.Errorf(groupRequired)
+	}
+	if len(opts.Version) == 0 {
+		return fmt.Errorf(versionRequired)
+	}
+	if len(opts.Kind) == 0 {
+		return fmt.Errorf(kindRequired)
+	}
+
+	// Check if the qualified group has a valid DNS1123 subdomain value
+	if err := validation.IsDNS1123Subdomain(opts.QualifiedGroup()); err != nil {
+		return fmt.Errorf("either group or domain is invalid: (%v)", err)
+	}
+
+	// Check if the version follows the valid pattern
+	if !versionRegex.MatchString(opts.Version) {
+		return fmt.Errorf("version must match %s (was %s)", versionPattern, opts.Version)
+	}
+
+	validationErrors := []string{}
+
+	// Require kind to start with an uppercase character
+	if string(opts.Kind[0]) == strings.ToLower(string(opts.Kind[0])) {
+		validationErrors = append(validationErrors, "kind must start with an uppercase character")
+	}
+
+	validationErrors = append(validationErrors, validation.IsDNS1035Label(strings.ToLower(opts.Kind))...)
+
+	if len(validationErrors) != 0 {
+		return fmt.Errorf("invalid Kind: %#v", validationErrors)
+	}
+
+	// TODO: validate plural strings if provided
+
+	// Ensure apiVersions for k8s types are empty or valid.
+	for typ, apiVersion := range map[string]string{
+		"CRD":     opts.CRDVersion,
+		"Webhook": opts.WebhookVersion,
+	} {
+		switch apiVersion {
+		case "", v1beta1, v1:
+		default:
+			return fmt.Errorf("%s version must be one of: v1, v1beta1", typ)
+		}
+	}
+
+	return nil
+}
+
+// QualifiedGroup returns the fully qualified group name with the available information.
+func (opts Options) QualifiedGroup() string {
+	if opts.Domain == "" {
+		return opts.Group
+	}
+	return fmt.Sprintf("%s.%s", opts.Group, opts.Domain)
+}
+
+// GVK returns the GVK identifier of a resource.
+func (opts Options) GVK(isProjectVersion2 bool) resource.GVK {
+	if isProjectVersion2 {
+		return resource.GVK{
+			Group:   opts.Group,
+			Version: opts.Version,
+			Kind:    opts.Kind,
+		}
+	}
+	return resource.GVK{
+		Group:   opts.QualifiedGroup(),
```

Comment:
It is wrong. It should not use the QualifiedGroup at all. GKV should not contain the domain info. 
Also, it has a big chance to impact the v2/ behaviour.

Reply:
Again not true. GVK includes the domain info: [proof](https://github.com/kubernetes/kubernetes/blob/5a17c7ac7406689fc592d02ca2ffdb7097761ea5/pkg/apis/authorization/register.go#L25-L28).

We weren't considering it in v2 project files but thats why there is a conditional above that uses the `Group` without the `Domain`.

Reply:
In the Kubebuilder and in the solution of this PR `resource.GVK` represents:

```
resources:
- group: cache
  kind: Memcached
  version: v1alpha1
```

We do not store group as a `qualified group` then, I do not agree that `resource.GKV` should return : 

```
resources:
- group: cache + domain
  kind: Memcached
  version: v1alpha1
```

c/c @estroz

Reply:
GVK is a k/k concept and it uses the group with the domain. The division into group and domain is something artificial that we made out to avoid repetition (prevent the user to have to specify the domain every command call). As we are gonna be storing not only internal resources (core builtin types are external for example) we need to use the real GVK as the key. Before, only local resources whwere stored in the config file so the partial key that we were using was enough to uniquely identify each resource but not any more, the domain is needed, adn thus is why the qualified group name is required.